### PR TITLE
Fix VRF central condenser EIRFPLR curve bound and add warning messages

### LINKED ISSUE #9572: VRF central condenser EIR curves not checked and bounded causing negative cooling electricity
Issue overview
--------------
It was discovered that in the curved based AirConditioner:VariableRefrigerantFlow simulation code, the EIR (as a function of PLR) curve bounds were not checked to ensure a >0 value. In this case, when a bad curve is entered, negative cooling electricity will be calculated without any warnings. 

It is suggested to catch this exception by: 
1.	giving with recurring warning messaged about the situation;
2.	reset the negative curve value to 0.0 before continuing the simulation.

The same situation is happening to the heating EIR in the code. Similar treatment is suggested for the heating part as well.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): All
 - Version of EnergyPlus (if using an intermediate build, include SHA): v22.1
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9572))
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #9572 
 - When a bad curve input is given, the AirConditioner:VariableRefrigerantFlow object could yield negative cooling electricity usage without any warning sign. The fix applies a bound (>=0.0) for the EIR curve values and at the same time issues recurring warnings about the error. 
 - The fix is also applied to a heating EIRFPLR curve handling that could potentially have the similar issues.  

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/HVACVariableRefrigerantFlow.hh [hunk 1]
```diff
@@ -297,12 +297,14 @@ namespace HVACVariableRefrigerantFlow {
...
-        int HeatCapFTErrorIndex; // warning message index
-        int CoolCapFTErrorIndex; // warning message index
-                                 // The following are for the Algorithm Type: VRF model based on physics, applicable for Fluid Temperature Control
-        int AlgorithmIUCtrl;     // VRF indoor unit contrl algorithm, 1-High sensible, 2-Te/Tc constant
+        Real64 VRFHeatRec;         // Heat Recovery heat reclaim power (W)
+        Real64 VRFHeatEnergyRec;   // Heat Recovery heat reclain energy (J)
+        int HeatCapFTErrorIndex;   // warning message index
+        int CoolCapFTErrorIndex;   // warning message index
+        int HeatEIRFPLRErrorIndex; // warning message index
+        int CoolEIRFPLRErrorIndex; // warning message index
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.hh]: The new trend is to initialize right here, that way you don't have to change the constructor. Mixing initialization methods is allowed (from @Myoldmopar).

int HeatCapFTErrorIndex = 0;

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 2]
```diff
@@ -1154,6 +1177,29 @@ void CalcVRFCondenser(EnergyPlusData &state, int const VRFCond)
...
+        if (EIRFPLRModFac < 0.0) {
+            if (vrf.HeatEIRFPLRErrorIndex == 0) {
+                ShowSevereMessage(state, fmt::format("{} \"{}\":", std::string(cVRFTypes(VRF_HeatPump)), vrf.Name));
+                ShowContinueError(state, fmt::format(" Heating EIR Modifier curve (function of PLR) output is negative ({:.3T}).", EIRFPLRModFac));
+                ShowContinueError(state,
+                                  fmt::format(" Negative value occurs using an outdoor air temperature of {:.1T} C and an average indoor air "
+                                              "wet-bulb temperature of {:.1T} C.",
+                                              CondInletTemp,
+                                              InletAirWetBulbC));
+                ShowContinueErrorTimeStamp(state, " Resetting curve output to zero and continuing simulation.");
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1182]: Cooling is OAT and IDWB, heating is indoor T and outdoor WB or DB, depending on outdoor temperature type. Check what the code does for heating to test the outdoor temperature type when evaluating performance curves.

    A19, \field Heating Performance Curve Outdoor Temperature Type
       \type choice
       \key DryBulbTemperature
       \key WetBulbTemperature

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 3]
```diff
@@ -1154,6 +1177,29 @@ void CalcVRFCondenser(EnergyPlusData &state, int const VRFCond)
...
+                state,
+                fmt::format("{} \"{}\": Heating EIR Modifier curve (function of PLR) output is negative warning continues...",
+                            PlantEquipTypeNames[static_cast<int>(PlantEquipmentType::HeatPumpVRF)],
+                            vrf.Name),
+                vrf.HeatEIRFPLRErrorIndex,
+                EIRFPLRModFac,
+                EIRFPLRModFac);
+            EIRFPLRModFac = 0.0;
+        }
+
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1194]: These warnings do look good.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.hh (hunk 1 above)
Comment (jcyuan2020) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.hh]: Thanks! Changed to this way for the error index variables that appeared in the current struct.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc (hunk 2 above)
Comment (jcyuan2020) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1182]: It turned out here probably a more proper message should be given the PLR number. So changed this message text and variables according to that.

Comment (rraustad) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1182]: That looks much better.

**Comment by rraustad** [comment]
Is the defect file running on your machine? I get this message when I run the defect file.

    ** Severe  ** AirConditioner:VariableRefrigerantFlow "CENTRAL VRF CONDENSER":
    ** Severe  ** invalid type specifier

Tracking it down it appears to be something to do with the new warning message format. It looks like it's trying to parse 'T'

image

**Comment by rraustad** [comment]
When I hit line 1144 and continue,

image

It hits this before hitting line 1145

image

**Comment by rraustad** [comment]
Probably something to do with the format for a negative number.

**Comment by jcyuan2020** [comment]
It looks like that "T" does not seems to be listed as syntax for fmt::format here: https://fmt.dev/latest/syntax.html
But is using the defined `format` (instead of `fmt::format`), there is a defintion of type T (Line 420 onward in IOFiles.hh): 
```
        } else if (specs_.type == 'T') { // matches TrimSigDigits behavior
            const auto fixed_output = should_be_fixed_output(val);

            if (fixed_output) {
                const auto magnitude = std::pow(10, specs_.precision);
                const auto adjusted = (val * magnitude) + 0.0001;
                const auto truncated = std::trunc(adjusted) / magnitude;
                specs_.type = 'F';
                return fmt::format_to(ctx.out(), spec_builder(), truncated);
            } else {
                specs_.type = 'E';
                specs_.precision += 2;

                // write the `E` formatted float to a std::string
                auto str = fmt::format(spec_builder(), val);
                str = zero_pad_exponent(str);

                // Erase last 2 numbers to truncate the value
                const auto E_itr = std::find(begin(str), end(str), 'E');
                if (E_itr != str.end()) {
                    str.erase(std::prev(E_itr, 2), E_itr);
                }

                return fmt::format_to(ctx.out(), "{}", str);
            }
        }
```

**Comment by jcyuan2020** [comment]
So probably need to use the `format` instead of `fmt::format` for format type `T`.

**Comment by jcyuan2020** [comment]
Changed to using `format` for the two with `T` specifiers and then the defect file runs fine. So that seems to be the reason.

**Comment by rraustad** [comment]
This looks good, and brings up another point. If the EIRfPLR curve results in negative numbers the energy consumption will be 0. That is certainly 1 way to reduce energy use but not indicative of real machines. So what could we do to help the user understand this better? Will the user even see this warning? Do we force a fatal in this situation? and require the user to evaluate this curve? This is out of scope for this change but certainly worth thinking about.

    ** Severe  ** AirConditioner:VariableRefrigerantFlow "CENTRAL VRF CONDENSER":
    **   ~~~   **  Cooling EIR Modifier curve (function of PLR) output is negative (-0.290).
    **   ~~~   **  Negative value occurs using a cooling Part Load Ratio (PLR) of 0.25.
    **   ~~~   **  Resetting curve output to zero and continuing simulation.
    **   ~~~   **  During Warmup, Environment=NEW YORK CENTRAL PRK OBS BELV COOLING DESIGN DAY 0.4% CONDNS DB=>MWB, at Simulation time=07/21 00:00 - 00:10

    *************  ** Warning ** AirConditioner:VariableRefrigerantFlow "CENTRAL VRF CONDENSER": Cooling EIR Modifier curve (function of PLR) output is negative warning continues...
    *************  **   ~~~   **   This error occurred 47465 total times;
    *************  **   ~~~   **   during Warmup 7 times;
    *************  **   ~~~   **   during Sizing 0 times.
    *************  **   ~~~   **   Max=-1.581924E-003  Min=-0.290696

image

**Comment by jcyuan2020** [comment]
Probably one way is to add a couple of lines in the warning message to emphasize that it will produce false or inaccurate results (and probably false savings opportunities that does not exist in reality), and urge the user to check and correct the curve input. Actually set it to fatal might also be a feasible idea given it severity. 
There are a few other curve checks (e.g. the EITFT, CapFT etc in the same module that could apply the same principle.

**Review by rraustad — APPROVED** [review]
This adds a new warning for negative EIRfPLR curves for VRF. Warnings look good and address the issue.

**Comment by rraustad** [comment]
The warning itself is the emphasis. I think we should discuss this concept (free energy) on a future technicalities call.

**Comment by rraustad** [comment]
@jcyuan2020 could you set the milestone and update the status (currently draft).

File: src/EnergyPlus/HVACVariableRefrigerantFlow.hh [hunk 4]
```diff
@@ -389,27 +391,25 @@ namespace HVACVariableRefrigerantFlow {
...
+              C1Tc(0.0), C2Tc(0.0), C3Tc(0.0), DiffOUTeTo(5), EffCompInverter(0.95), EvaporatingTemp(6.0), EvapTempFixed(0.0), HROUHexRatio(0.0),
+              IUEvaporatingTemp(6.0), IUCondensingTemp(44.0), IUEvapTempLow(4.0), IUEvapTempHigh(15.0), IUCondTempLow(42.0), IUCondTempHigh(46.0),
+              IUCondHeatRate(0.0), IUEvapHeatRate(0.0), Ncomp(0.0), NcompCooling(0.0), NcompHeating(0.0), OUEvapTempLow(-30.0), OUEvapTempHigh(20.0),
+              OUCondTempLow(30.0), OUCondTempHigh(96.0), OUAirFlowRate(0.0), OUAirFlowRatePerCapcity(0.0), OUCondHeatRate(0.0), OUEvapHeatRate(0.0),
+              OUFanPower(0.0), RatedEvapCapacity(40000.0), RatedHeatCapacity(0.0), RatedCompPower(14000.0), RatedCompPowerPerCapcity(0.35),
+              RatedOUFanPower(0.0), RatedOUFanPowerPerCapcity(0.0), RateBFOUEvap(0.45581), RateBFOUCond(0.21900), RefPipDiaSuc(0.0),
+              RefPipDiaDis(0.0), RefPipLen(0.0), RefPipEquLen(0.0), RefPipHei(0.0), RefPipInsThi(0.0), RefPipInsCon(0.0), SH(0.0), SC(0.0), SCHE(0.0),
+              SHLow(0.0), SCLow(0.0), SHHigh(0.0), SCHigh(0.0), VRFOperationSimPath(0.0), checkPlantCondTypeOneTime(true)
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.hh:412]: 2 existing struct variables were initialized (e.g., int HeatCapFTErrorIndex = 0), changing the order of the constructor. That is perfectly acceptable, and encouraged.

File: tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc [hunk 5]
```diff
@@ -15932,4 +15933,245 @@ TEST_F(EnergyPlusFixture, VRFTU_FanOnOff_Power)
...
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalCoolLoad(1) = 0.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalCoolLoad(2) = 0.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalCoolLoad(3) = 0.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalCoolLoad(4) = 0.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalCoolLoad(5) = 0.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalHeatLoad(1) = 1000.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalHeatLoad(2) = 1000.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalHeatLoad(3) = 1000.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalHeatLoad(4) = 1000.0;
+    state->dataHVACVarRefFlow->TerminalUnitList(1).TotalHeatLoad(5) = 1000.0;
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:16143]: 1000 * 5 = 5000

File: tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc [hunk 6]
```diff
@@ -15932,4 +15933,245 @@ TEST_F(EnergyPlusFixture, VRFTU_FanOnOff_Power)
...
+
+    CalcVRFCondenser(*state, VRFCond);
+    EXPECT_FALSE(state->dataHVACVarRefFlow->VRF(VRFCond).HRHeatingActive);
+    EXPECT_FALSE(state->dataHVACVarRefFlow->VRF(VRFCond).HRCoolingActive);
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).TotalCoolingCapacity, 0.0);
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).TUCoolingLoad, 0.0);
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).TotalHeatingCapacity, 5000.0);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:16150]: and this model gives 5000. pretty impressive.

File: tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc [hunk 7]
```diff
@@ -15932,4 +15933,245 @@ TEST_F(EnergyPlusFixture, VRFTU_FanOnOff_Power)
...
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).VRFCondRTF, 1.0);
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).SUMultiplier, 1.0);
+    EXPECT_FALSE(state->dataHVACVarRefFlow->VRF(VRFCond).ModeChange);
+    EXPECT_FALSE(state->dataHVACVarRefFlow->VRF(VRFCond).HRModeChange);
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).ElecCoolingPower, 0.0);
+
+    // This should be reset to zero after the fix
+    EXPECT_NE(state->dataHVACVarRefFlow->VRF(VRFCond).ElecHeatingPower,
+              state->dataHVACVarRefFlow->VRF(VRFCond).RatedHeatingPower * state->dataHVACVarRefFlow->VRF(VRFCond).VRFCondPLR);
+    EXPECT_EQ(state->dataHVACVarRefFlow->VRF(VRFCond).ElecHeatingPower, 0.0);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:16162]: Please explain. The VRF just finished simulating. Why is the ElecHeatingPower = 0 if there actually was heating power? I would think heating power would be initialized to 0 in init, after all other reporting was complete and before then next iteration of the model.

File: tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc [hunk 8]
```diff
@@ -15932,4 +15933,245 @@ TEST_F(EnergyPlusFixture, VRFTU_FanOnOff_Power)
...
+    EXPECT_EQ(state->dataErrTracking->TotalWarningErrors, 2);
+    EXPECT_EQ(state->dataErrTracking->TotalSevereErrors, 0);
+    EXPECT_EQ(state->dataErrTracking->LastSevereError, "AirConditioner:VariableRefrigerantFlow \"\":");
+    EXPECT_EQ(state->dataErrTracking->NumRecurringErrors, 2);
+    EXPECT_EQ(state->dataErrTracking->RecurringErrors[1].MinValue, -0.25);
+    EXPECT_EQ(state->dataErrTracking->RecurringErrors[1].MaxValue, -0.25);
+    EXPECT_EQ(state->dataErrTracking->RecurringErrors[1].Count, 1);
+    EXPECT_EQ(state->dataErrTracking->RecurringErrors[1].Message,
+              " ** Warning ** AirConditioner:VariableRefrigerantFlow \"\": Heating EIR Modifier curve (function of PLR) output is negative warning "
+              "continues...");
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:16174]: tracking warnings, OK.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 9]
```diff
@@ -1137,6 +1137,25 @@ void CalcVRFCondenser(EnergyPlusData &state, int const VRFCond)
...
+                state,
+                fmt::format("{} \"{}\": Cooling EIR Modifier curve (function of PLR) output is negative warning continues...",
+                            PlantEquipTypeNames[static_cast<int>(PlantEquipmentType::HeatPumpVRF)],
+                            vrf.Name),
+                vrf.CoolEIRFPLRErrorIndex,
+                EIRFPLRModFac,
+                EIRFPLRModFac);
+            EIRFPLRModFac = 0.0;
+        }
+
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1158]: Cooling EIR negative EIRfPLR warning covered.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 10]
```diff
@@ -1154,6 +1173,25 @@ void CalcVRFCondenser(EnergyPlusData &state, int const VRFCond)
...
+            ShowRecurringWarningErrorAtEnd(
+                state,
+                fmt::format("{} \"{}\": Heating EIR Modifier curve (function of PLR) output is negative warning continues...",
+                            PlantEquipTypeNames[static_cast<int>(PlantEquipmentType::HeatPumpVRF)],
+                            vrf.Name),
+                vrf.HeatEIRFPLRErrorIndex,
+                EIRFPLRModFac,
+                EIRFPLRModFac);
+            EIRFPLRModFac = 0.0;
+        }
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1193]: Heating EIR negative EIRfPLR warning covered.

**Comment by rraustad** [comment]
Just 1 question to answer. Why is heating power = 0 after the VRF simulates heating power. PLR = 0.25 so there must be some power? But if the perf curves are bad then you would get 0 power. Hmm, I don't like that but that is what we do now. We have to figure something out here.

**Comment by rraustad** [comment]
OK, so this all looks good. I understand why we get 0 power with a bad EIRfPLR curve. Just don't like that answer but that is out of scope ... for now.

**Review by rraustad — APPROVED** [review]
These changes are approved with needed discussion of what to do with bad curves. For all models.

**Comment by rraustad** [comment]
@Myoldmopar CI is still running. This looks good to merge.

**Comment by Myoldmopar** [comment]
This is 100% clean, reviewed and approved, and ready to go.  Thanks @jcyuan2020 and reviewer crew.

**Pull request merged**